### PR TITLE
[POAE7-2483] ForEach and IF Control Flow Support

### PR DESCRIPTION
**Pull request by bigPYJ1151**
### What changes were proposed in this pull request?

Support control flow in JITLib 

### Why are the changes needed?

For nextgen.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UTs

### Which label does this PR belong to?

Infra

Pls take a review @xieqi @Ma-Jian1 @harborn @winningsix

### CODE REVIEW

File: cider/exec/nextgen/jitlib/base/JITControlFlow.h [hunk 1]
```diff
@@ -21,8 +21,32 @@
...
+class ForBuilder {
+ public:
+  virtual ~ForBuilder() = default;
+
+  virtual void build(const std::function<JITValuePointer()>& condition,
+                     const std::function<void()>& main_block,
+                     const std::function<void()>& update_block) = 0;
+};
+
+using IfBuilderPointer = std::unique_ptr<IfBuilder>;
```
Comment (chaojun-zhang) [inline cider/exec/nextgen/jitlib/base/JITControlFlow.h:48]: can we simply this pointer name as 'IfBuilderPtr' as same as we defining pointers elsewhere?

File: cider/exec/nextgen/jitlib/base/JITControlFlow.h [hunk 2]
```diff
@@ -21,8 +21,32 @@
...
+using ForBuilderPointer = std::unique_ptr<ForBuilder>;
```
Comment (chaojun-zhang) [inline cider/exec/nextgen/jitlib/base/JITControlFlow.h:49]: same as above

File: cider/exec/nextgen/jitlib/base/JITControlFlow.h [hunk 3]
```diff
@@ -21,8 +21,32 @@
...
+
 #include "exec/nextgen/jitlib/base/JITValue.h"
 
-namespace cider::jitlib {};  // namespace cider::jitlib
+namespace cider::jitlib {  // namespace cider::jitlib
+class IfBuilder {
+ public:
+  virtual ~IfBuilder() = default;
+
+  virtual void build(
```
Comment (chaojun-zhang) [inline cider/exec/nextgen/jitlib/base/JITControlFlow.h:33]: I guess you are try to introduce builder pattern here,  if so, can we separate these method parameters in their own setter method, and then have a build method with empty parameter to build the thing we want.

For example:
 Ifbuilder.condition(xx).if_true(xx).else_block(xx).build();

This library bytecode from airlist does the same thing, for more information see below link:
https://github.com/airlift/bytecode/blob/master/src/main/java/io/airlift/bytecode/control/IfStatement.java#L60
https://github.com/airlift/bytecode/blob/master/src/main/java/io/airlift/bytecode/control/IfStatement.java#L84

**Review by winningsix — APPROVED** [review]

File: cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp [hunk 4]
```diff
@@ -183,6 +184,14 @@ JITValuePointer LLVMJITFunction::getArgument(size_t index) {
                                             false);
   }
 }
+
+IfBuilderPointer LLVMJITFunction::getIfBuilder() {
+  return std::make_unique<LLVMIfBuilder>(func_, *ir_builder_);
+}
```
Comment (Ma-Jian1) [inline cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp:190]: `createIfBuilder` may be better than `getIfBuilder`, cause we always create a new instance.

File: cider/tests/nextgen/jitlib/JITLibTest.cpp [hunk 5]
```diff
@@ -143,6 +143,169 @@ TEST_F(JITLibTests, LogicalOpTest) {
...
+  module.finish();
+
+  auto func_ptr = function->getFunctionPointer<int32_t, int32_t, bool>();
+  EXPECT_EQ(func_ptr(123, false), 123);
+  EXPECT_EQ(func_ptr(123, true), 124);
+}
+
+TEST_F(JITLibTests, BasicIFControlFlowWithElseTest) {
+  LLVMJITModule module("TestModule");
+  JITFunctionPointer function = module.createJITFunction(JITFunctionDescriptor{
```
Comment (Ma-Jian1) [inline cider/tests/nextgen/jitlib/JITLibTest.cpp:175]: add pseudo code in the comment may be more clear?
```cpp
// func(int32 x, bool condition) {
//     if (condition) {
//         x = x + 1;
//     } else {
//         x = x + 10;
//     }
//     return x;
// }
```

File: cider/exec/nextgen/jitlib/llvmjit/LLVMJITControlFlow.cpp [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+static llvm::Value* castToBool(llvm::IRBuilder<>& builder, llvm::Value* value) {
+  if (value->getType()->isIntegerTy() && value->getType()->getIntegerBitWidth() != 1) {
+    return builder.CreateICmpNE(value, llvm::ConstantInt::get(value->getType(), 0));
+  } else {
+    return value;
+  }
+}
+
+void LLVMIfBuilder::build(const std::function<JITValuePointer()>& condition,
+                          const std::function<void()>& if_true_block,
```
Comment (jikunshang) [inline cider/exec/nextgen/jitlib/llvmjit/LLVMJITControlFlow.cpp:40]: will we provide some utils to create function here in the future? I am not clear how will a complex Expression(like NestedIFControlFlowTest) generate such function automatically.

File: cider/exec/nextgen/jitlib/base/JITControlFlow.h [hunk 7]
```diff
@@ -21,8 +21,32 @@
...
+};
+
+class ForBuilder {
+ public:
+  virtual ~ForBuilder() = default;
+
+  virtual void build(const std::function<JITValuePointer()>& condition,
+                     const std::function<void()>& main_block,
+                     const std::function<void()>& update_block) = 0;
+};
```
[the remark this answers is not in the document]
Comment (bigPYJ1151) [reply cider/exec/nextgen/jitlib/base/JITControlFlow.h:46]: I think so.

File: cider/exec/nextgen/jitlib/llvmjit/LLVMJITControlFlow.cpp (hunk 6 above)
Comment (bigPYJ1151) [reply cider/exec/nextgen/jitlib/llvmjit/LLVMJITControlFlow.cpp:40]: Multi-level nested if could be generated by a recursive function call if it has fixed pattern, just like ```NestedForControlFlowTest``` case shown.

File: cider/exec/nextgen/jitlib/base/JITControlFlow.h (hunk 1 above)
Comment (bigPYJ1151) [reply cider/exec/nextgen/jitlib/base/JITControlFlow.h:48]: Maybe we could do this after all JITLib features stable.

File: cider/tests/nextgen/jitlib/JITLibTest.cpp (hunk 5 above)
Comment (bigPYJ1151) [reply cider/tests/nextgen/jitlib/JITLibTest.cpp:176]: Seems the code structure is more clear after using pattern advised by chaojun.

File: cider/exec/nextgen/jitlib/llvmjit/LLVMJITControlFlow.cpp [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+static llvm::Value* castToBool(llvm::IRBuilder<>& builder, llvm::Value* value) {
+  if (value->getType()->isIntegerTy() && value->getType()->getIntegerBitWidth() != 1) {
+    return builder.CreateICmpNE(value, llvm::ConstantInt::get(value->getType(), 0));
+  } else {
+    return value;
+  }
+}
+
+void LLVMIfBuilder::build() {
+  if (condition_) {
```
Comment (Ma-Jian1) [inline cider/exec/nextgen/jitlib/llvmjit/LLVMJITControlFlow.cpp:41]: minimize the indentation
```cpp
if (condition_) {
    log(FATAL) <<
}
// do something else
```
or
```cpp
check(!condition_)
// do something else
```

File: cider/exec/nextgen/jitlib/llvmjit/LLVMJITControlFlow.cpp [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    builder_.CreateBr(after_bb);
+
+    builder_.SetInsertPoint(after_bb);
+  } else {
+    LOG(FATAL) << "Condition builder function is missed in LLVMIfBuilder::build.";
+  }
+}
+
+void LLVMLoopBuilder::build() {
+  if (condition_) {
```
Comment (Ma-Jian1) [inline cider/exec/nextgen/jitlib/llvmjit/LLVMJITControlFlow.cpp:76]: same with above

**Review by Ma-Jian1 — DISMISSED** [review]
LGTM

File: cider/exec/nextgen/jitlib/llvmjit/LLVMJITControlFlow.cpp [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+  if (if_true_) {
+    if_true_();
+  }
+  builder_.CreateBr(after_bb);
+
+  builder_.SetInsertPoint(false_bb);
+  if (if_false_) {
+    if_false_();
+  }
+  builder_.CreateBr(after_bb);
```
Comment (yma11) [inline cider/exec/nextgen/jitlib/llvmjit/LLVMJITControlFlow.cpp:67]: For this IfBuilder, the output code will be like `if(true){if_true_(); after_bb;}if(false){if_false_(); after_bb;}`? What will the `after_bb` do?

Comment (bigPYJ1151) [reply cider/exec/nextgen/jitlib/llvmjit/LLVMJITControlFlow.cpp:67]: Yes. Follow-up IR will be appended into ```after_bb```.

**Review by Ma-Jian1 — DISMISSED** [review]

File: cider/exec/nextgen/jitlib/llvmjit/LLVMJITControlFlow.cpp (hunk 10 above)
Comment (yma11) [reply cider/exec/nextgen/jitlib/llvmjit/LLVMJITControlFlow.cpp:67]: Thus we can move `after_bb` out of `if(true)` and `if(false)`, like `if(true){if_true_();} if(false){if_false_(); } after_bb;`

File: cider/tests/nextgen/jitlib/JITLibTest.cpp [hunk 11]
```diff
@@ -143,6 +143,176 @@ TEST_F(JITLibTests, LogicalOpTest) {
...
+  }
+  function->finish();
+  module.finish();
+
+  auto func_ptr = function->getFunctionPointer<int32_t, int32_t, bool>();
+  EXPECT_EQ(func_ptr(123, false), 133);
+  EXPECT_EQ(func_ptr(123, true), 124);
+}
+
+TEST_F(JITLibTests, NestedIFControlFlowTest) {
```
Comment (yma11) [inline cider/tests/nextgen/jitlib/JITLibTest.cpp:203]: What about multi `IFControlFlow` exist in single function?

Comment (bigPYJ1151) [reply cider/tests/nextgen/jitlib/JITLibTest.cpp:203]: Just create them sequentially.
```c++
auto if_builder1 = func->createIfBuilder();
auto if_builder2 = func->createIfBuilder();
...

if_builder1->condition()->...->build();
if_builder2->condition()->...->build();
...
```

File: cider/exec/nextgen/jitlib/llvmjit/LLVMJITControlFlow.cpp (hunk 10 above)
Comment (bigPYJ1151) [reply cider/exec/nextgen/jitlib/llvmjit/LLVMJITControlFlow.cpp:67]: Exactly.

**Review by Ma-Jian1 — APPROVED** [review]

**Review by jikunshang — APPROVED** [review]

**Pull request merged**